ibis/backends/sql/rewrites.py

```diff
@@ -244,6 +255,22 @@ def merge_select_select(_, **kwargs):
     if _.parent.find_below(blocking, filter=ops.Value):
         return _
 
+    if not _.distinct and _.parent.distinct:
+        # SELECT ... FROM (SELECT DISTINCT ...)
+        if all(isinstance(v, ops.Field) for v in _.selections.values()):
+            # Can fuse a trivial subselection after a distinct, like
+            #   SELECT a, b FROM (SELECT DISTINCT ...) -> SELECT DISTINCT a, b, ...
+            distinct = True
+        else:
+            # Other cases aren't safe to fuse as they might introduce non-distinct
+            # columns where applying a distinct would change the result.
+            return _
+    else:
+        # SELECT DISTINCT ... FROM (SELECT DISTINCT ...) -> SELECT DISTINCT ...
```

Comment:
What about this?

```sql
SELECT DISTINCT RANDOM_FUNCTION(a)
FROM (
  SELECT DISTINCT a -- assume this subquery returns one row
  FROM t
)
```

I don't think that's equivalent to

```sql
SELECT DISTINCT RANDOM_FUNCTION(a)
FROM t
```

Reply:
Hmmm, you're correct. We do balk already (above) if either the inner or outer query do contain impure functions like `random`. If you're ok with the logic remaining the same, I'll add a test for this case to ensure it doesn't change.

Reply:
Hmmm, given the above example though, I wonder if we don't want to rewrite it anyway. Generalizing this example to

```sql
SELECT DISTINCT SOME_FUNCTION(a)
FROM (
  SELECT DISTINCT a -- assume a has duplicates so this distinct reduces the set by a noticeable factor
  FROM t
)
```

We'd currently rewrite this to

```sql
SELECT DISTINCT SOME_FUNCTION(a) FROM t
```

If `SOME_FUNCTION` is deterministic, rewriting the former as the latter is correct.

However, if `SOME_FUNCTION` is expensive, doing the distinct in a subquery may be a major performance boost over the fused version (and rewriting the latter as the former isn't something I'd expect a backend's optimizer to do?).

Perhaps we never want to fuse chained distincts unless it's a trivial subselection, since fusing would prevent the user from doing such an optimization?

Reply:
> Perhaps we never want to fuse chained distincts unless it's a trivial subselection, since fusing would prevent the user from doing such an optimization?

This seems like a good first pass approach to me.